Use LRUCache for Hash Cache

danieldresser-ie opened the pull request:
I've already discussed this with John a fair bit.  Because of the need to make sure the hash is big enough to fit an entire root computation, the memory is still dependent on scene complexity, which we were hoping to fix.

However, I can now size the hash to fit 10 root computations, instead of 3200 root computations, and in my tests so far, this runs significantly faster.  ( Up to 2X ).  So this appears strictly better than before, and is probably worth using currently.

Also, the last commit with the early out in setMaxCost when cost is increased didn't actually seem necessary from timing with and without, I just didn't like the look of repeatedly calling limitCost and initializing a CacheEntry.

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -218,27 +224,46 @@ class ValuePlug::HashProcess : public Process
 		// computations, to prevent unbounded growth.
-		typedef std::pair<const ValuePlug *, IECore::MurmurHash> CacheKey;
 		typedef boost::unordered_map<CacheKey, IECore::MurmurHash> Cache;
+		*/
```
Could you just delete this old typedef rather than leave it hanging around?

johnhaddon commented on include/Gaffer/Private/IECorePreview/LRUCache.inl:
```diff
@@ -624,8 +624,15 @@ void LRUCache<Key, Value, Policy, GetterKey>::clear()
 {
-	m_maxCost = maxCost;
-	limitCost( maxCost );
+	if( maxCost > m_maxCost )
```
Could you make the equivalent PR for Cortex too please? I'd like to keep them in sync...

johnhaddon commented:
Nice one - thanks Daniel.

I should mention that in testing of the Cortex 10 hash cache vs the IECorePreview one replaced here, I did find cases where the Cortex 10 one is slower, because the parallel policy has to do more book-keeping to allow re-entrancy. That's one of the reasons I still haven't got the full-blown cache rewrite into final state, because I've been wondering if I can improve on it (or if the benefits outweigh the costs, and we can take a hit on simple graphs that were already fast to get big gains on more complex graphs). Here in this branch we won't be getting any of the benefits because we're not taking advantage of the re-entrancy. I'm talking purely about the parallel policy and the value cache here, in case that's not clear - as you say the serial policy for the hash cache is a clear win.

I'm not sure what to do about that. One option would be to add a non-reentrant version of the parallel policy and benchmark it to ensure there's no possibility of regression. Another would be to have _two_ LRUCaches here in IECorePreview - both the _new_ new one and the _old_ new one. A further option would be to just deploy this as-is, and treat it as an experiment in gathering more real-world performance data. I must admit I'm tempted by the latter, because it moves us forwards in terms of testing for the deployment of the full-blown cache rewrite, and because your performance numbers for the cases we care about the most right now are so good. What are your thoughts?

andrewkaufman commented:
I'm in favor of moving forward with the possible performance regression. @danieldresser-ie already tested that the current problem scenes in production are a clear win here, and we can't guarantee that the future holds anything better anyways, so it seems better to just get the improvements on the floor and live with the consequences, if they even get noticed on the simpler scenes.

johnhaddon commented:
OK, let's give it a bash.

danieldresser-ie commented:
I've removed the deprecated typedef, and added this PR to match in Cortex:
https://github.com/ImageEngine/cortex/pull/617

I guess the current thinking is we should just merge this and see how it goes?

johnhaddon commented:
> I guess the current thinking is we should just merge this and see how it goes?

I'm happy with that if you guys think the benefits outweigh the potential costs for real production networks. It makes it easier for me to progress with the full rewrite too.
